homeassistant/components/zha/core/gateway.py

```diff
@@ -231,3 +300,65 @@ def async_update_device(self, sender):
 
         device_entity = async_create_device_entity(zha_device)
         await self._component.async_add_entities([device_entity])
+
+        if is_new_join:
+            device_info = async_get_device_info(self._hass, zha_device)
+            async_dispatcher_send(
+                self._hass,
+                ZHA_GW_MSG,
+                {
+                    TYPE: DEVICE_FULL_INIT,
+                    DEVICE_INFO: device_info
+                }
+            )
+
+
+@callback
+def async_capture_log_levels():
+    """Capture current logger levels for ZHA."""
+    return {
+        BELLOWS: logging.getLogger(BELLOWS).getEffectiveLevel(),
+        ZHA: logging.getLogger(ZHA).getEffectiveLevel(),
+        ZIGPY: logging.getLogger(ZIGPY).getEffectiveLevel(),
+        ZIGPY_XBEE: logging.getLogger(ZIGPY_XBEE).getEffectiveLevel(),
+        ZIGPY_DECONZ: logging.getLogger(ZIGPY_DECONZ).getEffectiveLevel(),
+    }
+
+
+@callback
+def async_set_logger_levels(levels):
+    """Set logger levels for ZHA."""
+    logging.getLogger(BELLOWS).setLevel(levels[BELLOWS])
+    logging.getLogger(ZHA).setLevel(levels[ZHA])
+    logging.getLogger(ZIGPY).setLevel(levels[ZIGPY])
+    logging.getLogger(ZIGPY_XBEE).setLevel(levels[ZIGPY_XBEE])
+    logging.getLogger(ZIGPY_DECONZ).setLevel(levels[ZIGPY_DECONZ])
+
+
+class LogRelayHandler(logging.Handler):
+    """Log handler for error messages."""
+
+    def __init__(self, hass, gateway):
+        """Initialize a new LogErrorHandler."""
+        super().__init__()
+        self.hass = hass
+        self.gateway = gateway
+
+    def emit(self, record):
+        """Relay log message via dispatcher."""
+        if self.gateway.debug_enabled:
```

Comment:
is this if-check needed since you remove handler when you enable/disable debug mode?

Reply:
Nope, remnant from earlier. Will remove